pkg/kubelet/kubelet.go

```diff
@@ -1025,137 +1014,204 @@ func (kl *Kubelet) syncPod(pod *api.BoundPod, dockerContainers dockertools.Docke
+	recentContainers, err := dockertools.GetRecentDockerContainersWithNameAndUUID(kl.dockerClient, GetPodFullName(pod), uid, containerName)
+	if err != nil {
+		glog.Errorf("Error listing recent containers:%s", dockerContainerName)
+		// TODO(dawnchen): error handling here?
+	}
+	// set dead containers to unready state
+	for _, c := range recentContainers {
+		kl.readiness.remove(c.ID)
+	}
+
+	if len(recentContainers) > 0 {
+		if pod.Spec.RestartPolicy.Never != nil {
+			glog.Infof("Already ran container with name %s, do nothing",
+				dockerContainerName)
+			return false
+			
 		}
-		if count > 0 {
-			// Re-list everything, otherwise we'll think we're ok.
-			dockerContainers, err = dockertools.GetKubeletDockerContainers(kl.dockerClient, false)
-			if err != nil {
-				glog.Errorf("Error listing containers %#v", dockerContainers)
-				return err
+		if pod.Spec.RestartPolicy.OnFailure != nil {
+			// Check the exit code of last run
+			if recentContainers[0].State.ExitCode == 0 {
+				glog.Infof("Already successfully ran container with name %s, do nothing",
+					dockerContainerName)
+				return false
 			}
 		}
 	}
-	containersToKeep[podInfraContainerID] = empty{}
+	return true
+}
 
-	podVolumes, err := kl.mountExternalVolumes(pod)
-	if err != nil {
-		if ref != nil {
-			record.Eventf(ref, "failedMount",
-				"Unable to mount volumes for pod %q: %v", podFullName, err)
+func (kl *Kubelet) getPodInfraContainer(podFullName string, uid types.UID, dockerContainers dockertools.DockerContainers) (api.PodStatus, dockertools.DockerID, bool) {
+	if podInfraDockerContainer, found, _ := dockerContainers.FindPodContainer(podFullName, uid, dockertools.PodInfraContainerName); found {
+		podInfraContainerID := dockertools.DockerID(podInfraDockerContainer.ID)
+		// Somethings is running - we can get Status from the Docker.
+		podStatus, err := kl.GetPodStatus(podFullName, uid)
+		if err != nil {
+			glog.Errorf("Unable to get pod with name %q and uid %q info with error(%v)", podFullName, uid, err)
 		}
-		glog.Errorf("Unable to mount volumes for pod %q: %v; skipping pod", podFullName, err)
-		return err
+		return podStatus, podInfraContainerID, true
 	}
+	return api.PodStatus{}, "", false
+}
+
+func (kl *Kubelet) syncPod(pod *api.BoundPod, dockerContainers dockertools.DockerContainers) error {
+	podFullName := GetPodFullName(pod)
+	uid := pod.UID
+	containersToKeep := make(map[dockertools.DockerID]*api.Container)
+	containersToStart := make(map[*api.Container]empty)
+	var createPodInfraContainer = false
+
+	glog.Infof("Syncing Pod, podFullName: %q, uid: %q", podFullName, uid)
 
-	podStatus, err := kl.GetPodStatus(podFullName, uid)
+	err := kl.MakeDataDirs(uid)
 	if err != nil {
-		glog.Errorf("Unable to get pod with name %q and uid %q info with error(%v)", podFullName, uid, err)
+		return err
+	}
+	
+	podStatus, podInfraContainerID, found := kl.getPodInfraContainer(podFullName, uid, dockerContainers)
+	if found {
+		containersToKeep[podInfraContainerID] = nil
```

Comment:
I think we do. Otherwise we'll delete infra container when deleting everything we don't want to keep.